File: python/freq2.py
#!/usr/bin/env python3

"""
freq2.py: compute frequencies for a single variable in an
    LRECL data file. Also computes cumulative frequencies,
    percentages, and cumulative percentages.
    
Original Perl version by Kent Nassen (v1.2, 2/07/99)
Python version by Kent Nassen 2/24/2026

Usage:
    freq2.py [-h] [-c COLUMNS] filename
    
    -c#-#      Start-End column numbers of variable
    -c#        Single column variable at column #
    filename   LRECL data file filename
    -h         Show this help screen
    
    Example: freq2.py -c5-7 filename.dat
    
    (Column numbers in the file start at column 1)

Limitations:
    - Can only print the first 20 characters of a value,
      although can compute frequencies for longer values.
    - The order of values in the output is set by a character
      sort rather than a numeric sort (the program knows
      nothing of variable types like numeric or string).
    - If you give a column range outside the logical record
      length of the data, you will still get output, but all
      values will be null.
"""

import sys
import argparse
import re
from collections import defaultdict


def display_usage(prog_name):
    """Display usage message and exit."""
    print(f"""
   *** I need column locations!

{prog_name}: compute frequencies for a single variable in an
    LRECL data file. Also computes cumulative frequencies,
    percentages, and cumulative percentages.
    --Kent Nassen (v1.2, 2/07/99)

Usage: {prog_name} [-h] [-c[#-#][#] filename]
       -c#-#      Start-End column numbers of variable
       -c#        Single column variable at column #
       filename   LRECL data file filename
       -h         Show this help screen

       Example: {prog_name} -c5-7 fylename.dat

       (Column numbers in the file start at column 1)
""", file=sys.stderr)
    sys.exit(1)


def parse_columns(col_spec):
    """
    Parse column specification.
    Returns (start_col, end_col) tuple (1-based indexing).
    """
    if not col_spec:
        return None, None
    
    # Check for range format: #-#
    match = re.match(r'^(\d+)-(\d+)$', col_spec)
    if match:
        start_col = int(match.group(1))
        end_col = int(match.group(2))
        return start_col, end_col
    
    # Check for single column format: #
    match = re.match(r'^(\d+)$', col_spec)
    if match:
        col = int(match.group(1))
        return col, col
    
    return None, None


def print_header(page_num, loc, fname):
    """Print page header."""
    print()
    print(f"   Page {page_num:02d}")
    print(f"   Frequencies for the values in columns {loc}")
    print(f"   in the file \"{fname}\"")
    print()
    print("                                           Cumulative  Cumulative")
    print("               Value   Frequency  Percent   Frequency     Percent")
    print("              -------  ---------  -------  ----------  ----------")


def print_frequencies(freq_dict, filename, loc, pagesize=57):
    """Print formatted frequency table with pagination."""
    if not freq_dict:
        print("\nNo data found.")
        return
    
    total_cases = sum(freq_dict.values())
    cumulative_freq = 0
    page_num = 1
    line_count = 0
    
    # Print first page header
    print_header(page_num, loc, filename)
    line_count = 7  # Header uses 7 lines
    
    # Sort keys - try numeric sort first, fall back to string sort
    try:
        sorted_keys = sorted(freq_dict.keys(), key=lambda x: float(x) if x.strip() else float('inf'))
    except (ValueError, TypeError):
        sorted_keys = sorted(freq_dict.keys())
    
    for value in sorted_keys:
        # Check if we need a new page
        if line_count >= pagesize:
            page_num += 1
            print_header(page_num, loc, filename)
            line_count = 7
        
        freq = freq_dict[value]
        cumulative_freq += freq
        
        # Calculate percentages
        if total_cases > 0:
            freq_pct = (freq / total_cases) * 100
            cumulative_pct = (cumulative_freq / total_cases) * 100
        else:
            freq_pct = 0
            cumulative_pct = 0
        
        # Truncate value to 20 characters for display
        display_value = value[:20] if len(value) > 20 else value
        
        # Print formatted line
        print(f"{display_value:>20} {freq:>9}  {freq_pct:>7.2f}  {cumulative_freq:>10}     {cumulative_pct:>7.2f}")
        line_count += 1
    
    print()


def main():
    """Main function."""
    prog_name = sys.argv[0].split('/')[-1].split('\\')[-1]
    
    # Custom argument parser to handle the specific format
    parser = argparse.ArgumentParser(
        add_help=False,
        description='Compute frequencies for LRECL data files'
    )
    parser.add_argument('-h', '--help', action='store_true',
                        help='Show help message')
    parser.add_argument('-c', dest='columns', type=str,
                        help='Column specification (e.g., 5-7 or 5)')
    parser.add_argument('filename', nargs='?',
                        help='LRECL data file to process')
    
    args = parser.parse_args()
    
    # Handle help
    if args.help or not args.columns:
        display_usage(prog_name)
    
    # Parse column specification
    start_col, end_col = parse_columns(args.columns)
    
    if start_col is None or end_col is None:
        print("\n   *** I couldn't understand your column location(s)!\n", file=sys.stderr)
        display_usage(prog_name)
    
    # Validate column numbers
    if start_col == 0 or end_col == 0:
        print("\n   *** Starting or ending column was zero.\n", file=sys.stderr)
        display_usage(prog_name)
    
    # Swap if start > end
    if start_col > end_col:
        start_col, end_col = end_col, start_col
        print("\n   *** Start col greater than end col: assuming reversed col. locations.\n",
              file=sys.stderr)
    
    # Check for filename
    if not args.filename:
        print("\n   *** I need a data filename!\n", file=sys.stderr)
        display_usage(prog_name)
    
    # Create location string for display
    loc = f"{start_col}-{end_col}"
    
    # Read data file and extract column values
    freq_dict = defaultdict(int)
    
    try:
        with open(args.filename, 'r', encoding='utf-8', errors='replace') as f:
            for line in f:
                line = line.rstrip('\n\r')
                # Extract substring (convert to 0-based indexing)
                value = line[start_col-1:end_col] if len(line) >= start_col else ''
                freq_dict[value] += 1
    except IOError as e:
        print(f"\n{prog_name}:   *** Can't open data file '{args.filename}': {e}\n",
              file=sys.stderr)
        sys.exit(1)
    
    # Print frequency table
    print_frequencies(freq_dict, args.filename, loc)


if __name__ == '__main__':
    main()

File: perl/freq2.pl
#!/usr/bin/perl
##---------------------------------------------------------------------------##
##  File:
##      @(#)  freq2.pl 1.2 99/02/07 @(#)
##  Author:
##      Kent Nassen     knassen@umich.edu
##  Description:
##      A program for grabbing quick and dirty univariate frequencies
##      from an LRECL data file (vars in fixed column locations). Useful
##      if you just want to see what's in a column range without having
##      to fire up SAS or SPSS.
##
##  Usage:
##      freq.pl [-h] [-c[#-#][#] filename]
##   where: -c#-# indicates starting & ending column numbers of the variable
##          and -c# indicates a single-column variable at column #.
##          -h gives a help screen.
##          Only one filename may be given.
##
##  Limitations:
##        - Can only print the first 20 characters of a value,
##          although can compute frequencies for longer values.
##        - The order of values in the output is set by a character
##          sort rather than a numeric sort (the program knows
##          nothing of variable types like numeric or string).
##        - If you give a column range outside the logical record
##          length of the data, you will still get output, but all
##          values will be null.
##
##  v1.1: Added computation of percentages (7/4/1997)
##  v1.2: Checked with -w and 'use strict'.  Improved error handling.
##        Fixed number of lines per page in output. (2/7/1999)
##      
##---------------------------------------------------------------------------##
#use warnings;
#use strict;

use Getopt::Std;

our ($opt_h, $ProgName, $opt_c, $temp, $loc, $fname, 
     $nofcases, $startcol, $endcol, $cfreq, $value, 
     $cfreqpct, $valuepct, $freqpct, @data, %freq,
     $transpos, $version, $pagesize);

$pagesize=57;
STDOUT->format_lines_per_page($pagesize);

# Check performance
#my $begint = (times) [0];

$version="(v1.2, 2/07/99)";

#$opt_h="";

($ProgName = $0) =~ s%.*/%%;  # Unix
#($ProgName = lc $0) =~ s%.*\\%%;  # DOS

# Usage message subroutine
sub DisplayUsage {
    print STDERR "\n$ProgName: compute frequencies for a single variable in an\n",
      "    LRECL data file. Also computes cumulative frequencies\,\n",
      "    percentages\, and cumulative percentages.\n",
      "    --Kent Nassen $version\n",
      "\nUsage: $ProgName [-h] [-c[#-#][#] filename]\n",
	  "       -c#-#      Start-End column numbers of variable\n",
	  "       -c#        Single column variable at column #\n",
  	  "       filename   LRECL data file filename\n",
	  "       -h         Show this help screen\n",
	  "\n",
	  "       Example: $ProgName -c5-7 fylename.dat\n",
	  "\n       (Column numbers in the file start at column 1)\n\n";
      exit 1;
}

# Get possible options (-c requires a parameter)
getopts (('hc:') or &DisplayUsage); 

# -h option?
if ($opt_h) { &DisplayUsage };

# Parse the -c option for starting and ending column locations
#  (ending location need not be present if the variable is only
#  one column wide, e.g. -c5)
if (!$opt_c) { print STDERR "\n   *** I need column locations!\n"; 
			&DisplayUsage;
			} # if we don't have an opt_c, no need to go further
$_ = $opt_c;
if (/^([0-9]+)-([0-9]+$)/) { # both starting and ending locations: #-#
	$startcol=$1; $endcol=$2;
	}
elsif (/^[0-9]+$/) { # just starting location: #
	$startcol=$opt_c; $endcol=$startcol;
	}
else { print STDERR "\n   *** I couldn't understand your column location(s)!\n"; 
	&DisplayUsage;
	}

# See if the column locations of the variable make sense
if ($endcol==0 || $startcol==0) {
	print STDERR "\n   *** Starting or ending column was zero.\n";
	&DisplayUsage;
	}

# Assume startcol > endcol is a transposition of column locations
# You may want to make this an error instead.
if ($startcol > $endcol) { # swap startcol and endcol
	$temp=$endcol; $endcol=$startcol; $startcol=$temp; 
    print STDERR "\n   *** Start col greater than end col: assuming reversed col. locations.\n";
    }

# Set up a string var for printing out hyphenated column locations 
# (rather than construct this for each page printed)
$loc="$startcol"."-"."$endcol";

# What is left on the command is the filename to be processed (hopefully)
if ($#ARGV>0) { print STDERR "\n   *** One filename at a time, please!\n";
		&DisplayUsage;
		}
if ($#ARGV<0) { print STDERR "\n   *** I need a data filename!\n";
		&DisplayUsage;
		}
open INFILE, "<$ARGV[0]" or die "\n$ProgName:   *** Can't open data file '$ARGV[0]': $!\n\n";

# Set up a string var for printing out a quoted filename 
#  (rather than construct this for each page printed)
$fname="\"".$ARGV[0]."\"";

#### ---------------------------------------------------- ####
#### Read the data from the file named on the commandline ####
#### ---------------------------------------------------- ####
while (<ARGV>) { chop;   
	# Pull the desired columns and save in data array
	push @data, substr($_,$startcol-1,$endcol-$startcol+1); 
	}

#### ---------------------------------------------------- ####
#### The real work begins here                            ####
#### ---------------------------------------------------- ####
# Accumulate Frequencies 
$nofcases=$cfreq=0;
foreach $value (@data) { $freq{$value}++; $nofcases++; }

# Print Out the Frequencies
foreach $value (sort { $a <=> $b } keys(%freq)) {
	$cfreq=$cfreq+$freq{$value}; # cumulative freqs
	if ($nofcases) { $cfreqpct=($cfreq/$nofcases)*100 } # cumulative percent
		else { $cfreqpct="----" } # don't divide by zero
	if ($nofcases) {$freqpct=($freq{$value}/$nofcases)*100;} # value percent
		else { $freqpct="----" }
	write; # this will be formatted by the formats at the end
	}
print "\n";
close INFILE;

# Check performance
#my $endt = (times)[0];
#printf("\n Preceding task required %.2f seconds CPU time\n",$endt-$begint);

# Formatted, paged output
format STDOUT_TOP =
 
   @||||@||
   "Page",$%
   @<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<@<<<<<<<<<<<<
   "Frequencies for the values in columns",$loc
   @<<<<<<<<<<<@<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
   "in the file ",$fname
 
                                           Cumulative  Cumulative
               Value   Frequency  Percent   Frequency     Percent
              -------  ---------  -------  ----------  ----------
.
format STDOUT =
@>>>>>>>>>>>>>>>>>>> @>>>>>>>>>  @###.##  @>>>>>>>>>     @###.##
$value,$freq{$value},$freqpct,$cfreq,$cfreqpct 
.
